### PR TITLE
feat: add update bucket info precompile

### PR DESCRIPTION
**Pull request by phenix3443**

### CODE REVIEW

File: solidity/contracts/storage/IStorage.sol [hunk 1]
```diff
@@ -328,6 +338,16 @@ interface IStorage {
         uint256 id
     );
 
+    /**
+     * @dev UpdateBucket defines an Event emitted when a user update a bucket
+     */
+    event CreateBucket(
+        address indexed operator,
+        string indexed bucketName,
```
[the remark this answers is not in the document]
Comment (athaswang2000) [reply solidity/contracts/storage/IStorage.sol:346]: CreateBucket  -> UpdateBucket

File: x/evm/precompiles/bank/IBank.go [hunk 2]
```diff
@@ -141,11 +140,11 @@ func NewIBankFilterer(address common.Address, filterer bind.ContractFilterer) (*
...
 // bindIBank binds a generic wrapper to an already deployed contract.
 func bindIBank(address common.Address, caller bind.ContractCaller, transactor bind.ContractTransactor, filterer bind.ContractFilterer) (*bind.BoundContract, error) {
-	parsed, err := IBankMetaData.GetAbi()
+	parsed, err := abi.JSON(strings.NewReader(IBankABI))
 	if err != nil {
 		return nil, err
 	}
-	return bind.NewBoundContract(address, *parsed, caller, transactor, filterer), nil
+	return bind.NewBoundContract(address, parsed, caller, transactor, filterer), nil
 }
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/bank/IBank.go]: Different abigen version changed this file.

File: x/evm/precompiles/storage/tx.go [hunk 3]
```diff
@@ -106,6 +109,40 @@ func (c *Contract) CreateBucket(ctx sdk.Context, evm *vm.EVM, contract *vm.Contr
...
+	if _,err:=server.UpdateBucketInfo(ctx,msg); err!=nil {
+		return nil,err
+	}
+	if err:=c.AddLog(evm, MustEvent(UpdateBucketInfoEventName),[]common.Hash{
+		common.BytesToHash(contract.Caller().Bytes()),
+		common.BytesToHash([]byte(args.BucketName)),
+		common.BytesToHash(args.PaymentAddress.Bytes()),
+	},args.Visibility)	;err!=nil {
+		return nil,err
+	}
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/storage/tx.go]: Fixed

File: x/evm/precompiles/virtualgroup/IVirtualGroup.go
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/virtualgroup/IVirtualGroup.go:1]: Different abigen version caused this problem.

File: x/evm/precompiles/storage/types.go [hunk 4]
```diff
@@ -81,6 +81,18 @@ func (args *CreateBucketArgs) Validate() error {
...
+type UpdateBucketArgs struct {
+	BucketName        string         `abi:"bucketName"`
+	ChargedReadQuota  uint64         `abi:"chargedReadQuota"`
+	PaymentAddress    common.Address `abi:"paymentAddress"`
+	Visibility        uint8          `abi:"visibility"`
+}
+
+// Validate CreateBucketArgs args
+func (args *UpdateBucketArgs) Validate() error {
+	return nil
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/storage/types.go:105]: Added check.

File: x/evm/precompiles/storage/tx.go [hunk 5]
```diff
@@ -106,6 +109,40 @@ func (c *Contract) CreateBucket(ctx sdk.Context, evm *vm.EVM, contract *vm.Contr
...
+	var args UpdateBucketArgs
+	if err:= types.ParseMethodArgs(method, &args, contract.Input[4:]); err != nil {
+		return nil,err
+	}
+	msg:=&storagetypes.MsgUpdateBucketInfo{
+		Operator: contract.CallerAddress.String(),
+		BucketName: args.BucketName,
+		Visibility: storagetypes.VisibilityType(args.Visibility),
+		PaymentAddress: args.PaymentAddress.String(),
+		ChargedReadQuota:&mechaincommon.UInt64Value{Value: args.ChargedReadQuota},
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/storage/tx.go]: Use a mask to special field to update.

File: solidity/contracts/storage/IStorage.sol [hunk 6]
```diff
@@ -181,6 +181,18 @@ interface IStorage {
...
+    /**
+     * @dev updateBucketInfo defines a method for update a bucket.
+     */
+    function updateBucketInfo(
+        string memory bucketName,
+        VisibilityType visibility,
+        address paymentAddress,
+        uint64 chargedReadQuota,
+        uint8 updateMask // 1: visibility, 2: paymentAddress, 4: chargedReadQuota
+    ) external returns (bool success);
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply solidity/contracts/storage/IStorage.sol]: Fixed

File: solidity/contracts/storage/IStorage.sol [hunk 7]
```diff
@@ -181,6 +181,18 @@ interface IStorage {
...
+
+    /**
+     * @dev updateBucketInfo defines a method for update a bucket.
+     */
+    function updateBucketInfo(
+        string memory bucketName,
+        VisibilityType visibility,
+        address paymentAddress,
+        uint64 chargedReadQuota,
+        uint8 updateMask // 1: visibility, 2: paymentAddress, 4: chargedReadQuota
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply solidity/contracts/storage/IStorage.sol]: Fixed

File: x/evm/precompiles/compile copy.sh
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/compile copy.sh:1]: Fixed

File: x/evm/precompiles/storage/tx.go [hunk 8]
```diff
@@ -106,6 +110,48 @@ func (c *Contract) CreateBucket(ctx sdk.Context, evm *vm.EVM, contract *vm.Contr
...
+		return nil, err
+	}
+	server := storagekeeper.NewMsgServerImpl(c.storageKeeper)
+	if _, err := server.UpdateBucketInfo(ctx, msg); err != nil {
+		return nil, err
+	}
+	bucketNameHash := crypto.Keccak256([]byte(args.BucketName))
+	if err := c.AddLog(evm, MustEvent(UpdateBucketInfoEventName), []common.Hash{
+		common.BytesToHash(contract.Caller().Bytes()),
+		common.BytesToHash(bucketNameHash),
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/storage/tx.go:155]: But more often we need to filter events by bucketName.

File: solidity/contracts/storage/IStorage.sol [hunk 9]
```diff
@@ -328,6 +340,16 @@ interface IStorage {
         uint256 id
     );
 
+    /**
+     * @dev UpdateBucketInfo defines an Event emitted when a user update a bucket
+     */
+    event UpdateBucketInfo(
+        address indexed operator,
+        bytes32 indexed bucketName,
+        address indexed PaymentAddress,
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply solidity/contracts/storage/IStorage.sol:348]: Fixed

File: x/evm/precompiles/storage/tx.go [hunk 10]
```diff
@@ -106,6 +110,48 @@ func (c *Contract) CreateBucket(ctx sdk.Context, evm *vm.EVM, contract *vm.Contr
...
+	}
+	if args.UpdateMask&uint8(UpdateBucketInfoMaskVisibility) != 0 {
+		msg.Visibility = storagetypes.VisibilityType(args.Visibility)
+	}
+	if args.UpdateMask&uint8(UpdateBucketInfoMaskPaymentAddress) != 0 {
+		msg.PaymentAddress = args.PaymentAddress.String()
+	}
+	if args.UpdateMask&uint8(UpdateBucketInfoMaskChargedReadQuota) != 0 {
+		msg.ChargedReadQuota = &mechaincommon.UInt64Value{Value: uint64(args.ChargedReadQuota)}
+	}
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/storage/tx.go]: Fixed

File: solidity/contracts/storage/IStorage.sol [hunk 11]
```diff
@@ -328,6 +340,16 @@ interface IStorage {
...
+        uint8 Visibility
+    );
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply solidity/contracts/storage/IStorage.sol]: But more often we need to filter events by bucketName.

File: solidity/contracts/storage/IStorage.sol [hunk 12]
```diff
@@ -328,6 +339,16 @@ interface IStorage {
...
     );
 
+    /**
+     * @dev UpdateBucketInfo defines an Event emitted when a user update a bucket
+     */
+    event UpdateBucketInfo(
+        address indexed operator,
+        bytes32 indexed bucketName,
+        address indexed paymentAddress,
+        int8 Visibility
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply solidity/contracts/storage/IStorage.sol:349]: Fixed

File: solidity/contracts/storage/IStorage.sol [hunk 13]
```diff
@@ -181,6 +181,17 @@ interface IStorage {
         uint64 chargedReadQuota
     ) external returns (bool success);
 
+
+    /**
+     * @dev updateBucketInfo defines a method for update a bucket.
+     */
+    function updateBucketInfo(
+        string memory bucketName,
+        int8 visibility,
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply solidity/contracts/storage/IStorage.sol:190]: Fixed

File: x/evm/precompiles/storage/tx.go [hunk 14]
```diff
@@ -106,6 +110,48 @@ func (c *Contract) CreateBucket(ctx sdk.Context, evm *vm.EVM, contract *vm.Contr
...
+	var args UpdateBucketInfoArgs
+	if err := types.ParseMethodArgs(method, &args, contract.Input[4:]); err != nil {
+		return nil, err
+	}
+	
+	msg := &storagetypes.MsgUpdateBucketInfo{
+		Operator:   contract.CallerAddress.String(),
+		BucketName: args.BucketName,
+		Visibility: storagetypes.VisibilityType(args.Visibility),
+		PaymentAddress: args.PaymentAddress.String(),
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/evm/precompiles/storage/tx.go]: Fixed

File: x/storage/keeper/keeper.go [hunk 15]
```diff
@@ -423,7 +423,9 @@ func (k Keeper) UpdateBucketInfo(ctx sdk.Context, operator sdk.AccAddress, bucke
 		// save quota update time
 		k.setQuotaUpdateTime(ctx, bucketInfo.Id, blockTime)
 	}
-
+	if opts.Visibility > types.VISIBILITY_TYPE_INHERIT {
+		return types.ErrInvalidVisibility
+	}
```
[the remark this answers is not in the document]
Comment (phenix3443) [reply x/storage/keeper/keeper.go]: Fiexed

**Pull request merged**